content/blog/2016/2016-06-16-parallel-test-executor-plugin.adoc

```diff
@@ -0,0 +1,191 @@
+machine even while other lab machines sit idle.  In these cases, parallelizing
+test execution across multiple machines is a great way to speed up pipelines.
+The Parallel Test Executor plugin lets us leverage Jenkins do just that with no
+disruption to the rest of the build system.
+
+
+=== Serial Test Execution
+
+For this post, I’ll be running a pipeline based on the
+link:https://github.com/jenkinsci/git-plugin[Jenkins Git Plugin]. I've modified
+the pipeline from that project to allow us to compare execution times to our
+later changes, and I've truncate the "mvn" utility method since it remains
+unchanged.  You can find the original file
+link:https://github.com/bitwiseman/git-plugin/blob/7a31858e61d2ca2e752b0e4f1285bddcb7a75c4d/Jenkinsfile[here].
+
+[source,groovy]
+----
+node {
+  stage 'Checkout'
+  checkout scm
+
+  stage 'Build'
+
+  /* Call the maven build without tests. */
+  mvn "clean install -DskipTests"
+
+  stage 'Test'
+  runTests()
+
+  /* Save Results. */
+  stage 'Results'
+
+  /* Archive the build artifacts */
+  step([$class: 'ArtifactArchiver', artifacts: 'target/*.hpi,target/*.jpi'])
+}
+
+void runTests(def args) {
+  /* Call the maven build with tests. */
+  mvn "install -Dmaven.test.failure.ignore=true"
+
+  /* Archive the test results */
+  step([$class: 'JUnitResultArchiver', testResults: '**/target/surefire-reports/TEST-*.xml'])
+}
+
+/* Run maven */
+void mvn(def args) { /* ... */ }
+----
+
+It’s a maven project, so the the pipeline is pretty simple.
+As noted above, I’ve split the maven build into separate “Build” and “Test”
+stages.  Maven doesn’t support this split very well, it wants to run its all
+the steps of the lifecycle in order every time. So, I have to call maven twice:
+first using the “skipTests” property to do only build steps in the first call,
+and then a second time with out that property to run tests.
+
+On my quad-core machine, executing this pipeline takes about 13 minutes and 30
+seconds.  Of that time, it takes 13 minutes to run about 2.7 thousand tests in
+serial.
+
+image::/images/post-images/2016-06-16/serial.png[Serial Test Pipeline, role=center]
+
+
+=== Parallel Test Execution
+
+This looks like an ideal project for parallel test execution.  A short build
+followed by a large number of serially executed tests that consume the most of
+the pipeline time.  There are a number of things I could try to optimize this.
+For example, I could modify test harness to look for ways to parallelize
+the test execution on this single machine.  Or I could try speed up the tests
+themselves.  Both of those can be time consuming and risk destablizing the
+tests. I'd need to know more about the project to do it well.
+
+I'll avoid that risk by using Jenkins and the
+link:https://wiki.jenkins-ci.org/display/JENKINS/Parallel+Test+Executor+Plugin[Parallel Test Executor Plugin] to
+parallelize the tests across multiple machines instead.  That way all I need to
```

Comment:
From my understand, this is going to parallelize tests across multiple *executors* not necessarily machines, correct?